lib/matrix_reloaded/matrix.ex

```diff
@@ -0,0 +1,1005 @@
+      iex> MatrixReloaded.Matrix.and_then2(mat1, mat2, &MatrixReloaded.Matrix.concat_row(&1, &2))
+      {:ok,
+        [
+          [1, 0, 0, 2, 0, 0],
+          [0, 1, 0, 0, 2, 0],
+          [0, 0, 1, 0, 0, 2]
+        ]
+      }
+
+  """
+  @spec concat_row(t(), t()) :: Result.t(String.t(), t())
+  def concat_row(matrix1, matrix2) do
+    {rs1, _cs1} = size(matrix1)
+    {rs2, _cs2} = size(matrix2)
+
+    if rs1 == rs2 do
+      matrix1
+      |> Enum.zip(matrix2)
+      |> Enum.map(fn {r1, r2} -> Enum.concat(r1, r2) end)
+      |> Result.ok()
+    else
+      Result.error("Matrices have different row dimensions. Must be same!")
+    end
+  end
+
+  @doc """
+  Concatenate matrices vertically. Both matrices must have same columns dimension.
+
+  Returns result, it means either tuple of `{:ok, matrix}` or `{:error, "msg"}`.
+
+  ##  Example:
+      iex> mat1 = MatrixReloaded.Matrix.diag([1, 1, 1])
+      iex> mat2 = MatrixReloaded.Matrix.diag([2, 2, 2])
+      iex> MatrixReloaded.Matrix.and_then2(mat1, mat2, &MatrixReloaded.Matrix.concat_col(&1, &2))
+      {:ok,
+        [
+          [1, 0, 0],
+          [0, 1, 0],
+          [0, 0, 1],
+          [2, 0, 0],
+          [0, 2, 0],
+          [0, 0, 2]
+        ]
+      }
+
+  """
+  @spec concat_col(t(), t()) :: Result.t(String.t(), t())
+  def concat_col(matrix1, matrix2) do
+    {_rs1, cs1} = size(matrix1)
+    {_rs2, cs2} = size(matrix2)
+
+    if cs1 == cs2 do
+      matrix1
+      |> Enum.concat(matrix2)
+      |> Result.ok()
+    else
+      Result.error("Matrices have different column dimensions. Must be same!")
+    end
+  end
+
+  @doc """
+  The size (dimensions) of the matrix.
+
+  Returns tuple of {row_size, col_size}.
+
+  ## Example:
+
+      iex> MatrixReloaded.Matrix.new({3,4}) |> Result.and_then(&MatrixReloaded.Matrix.size(&1))
+      {3, 4}
+
+  """
+  @spec size(t()) :: {pos_integer, pos_integer}
+  def size(matrix), do: {length(matrix), length(List.first(matrix))}
+
+  @doc """
+  Save a matrix to csv file at a project root directory. Default name
+  of csv file is `matrix.csv`. Or you can set a path to save dir and file name
+  (e.g. `/tmp/matrix.csv`).
+
+  Returns result, it means either tuple of {:ok, :ok} or {:error, "msg"}.
+
+  ## Example:
+
+      iex> mat_ones = MatrixReloaded.Matrix.new(3, 1)
+      iex> mat_zeros = MatrixReloaded.Matrix.new(5)
+      iex> mat = MatrixReloaded.Matrix.new(7, 1)
+      iex> mat = MatrixReloaded.Matrix.and_then2(mat, mat_zeros, &MatrixReloaded.Matrix.update(&1, &2, {1, 1}))
+      iex> mat = MatrixReloaded.Matrix.and_then2(mat, mat_ones, &MatrixReloaded.Matrix.update(&1, &2, {2, 2}))
+      iex> mat |> Result.and_then(&MatrixReloaded.Matrix.save_csv(&1))
+      {:ok, :ok}
+
+      iex> MatrixReloaded.Matrix.new({3,4}) |> Result.and_then(&MatrixReloaded.Matrix.save_csv(&1, "/tmp/matrix.csv"))
+      {:ok, :ok}
+
+  """
+  @spec save_csv(t(), String.t()) :: Result.t(String.t(), :ok)
```

Comment:
Pls, move this function to own module (eg. `MatrixReloaded.Csv`). Is it necessary to write to disk? For Inspection, it will be better to implement the [Inspect protocol](https://hexdocs.pm/elixir/Inspect.html#content).

Reply:
I'm not sure if it's worth to make new file for only one function. I think that `IO.inspect` is ok for "low" dimensions of matrices. But for bigger is reasonable to save into the file.